### ISSUE TITLE
Feature Request: add rsync as tool

### ISSUE THREAD

**Issue by C-Wx** [git-for-windows/git#347]
There was already a discussion on msysgit, to add rsync to the set of preinstalled tools. But sadly it never made it.

Why I suggest to add rsync:
I personaly think it is the better way to deploy my projects to production with rsync and not with git,
because it's more easy to exclude /.git and other unwanted items. You all have heard of public /.git folders in production environments with sensitve data. Whoever would ever have done this?
So I would recommend to spread more the way to deploy with rsync.

And of course if you have rsync already running on your windows machine you can use it for a lot more cool things.

And guys come on, the size of rsync is not that heavy...

**Comment by dscho** [comment]
@C-Wx if you really want it, can I see the well-crafted Pull Request?

**Comment by C-Wx** [comment]
If my skills would allow it to start with that feature myself, I wouldn't have come up with this question to you guys. So this was just input from an user to you guys to consider my suggestion.

I'm sorry for not being able to contribute more than this.

**Comment by dscho** [comment]
It's not exactly hard. See for yourself what others did: https://github.com/git-for-windows/build-extra/pull/57 (The only other resource you need is here: https://github.com/git-for-windows/git/wiki/Making-an-installer).

**Comment by joeyh** [comment]
My life would also be made a lot easier if rsync was included.

I think the best reason is just that git-fetch can use rsync. So, as ssh is included to handle ssh:// repos, it makes sense to include rsync for rsync:// repos

My personal reason is that git-annex uses rsync somewhat more heavily than git itself does, and git-annex for windows has a bear of a time including rsync and making it work with the ssh provided by git for windows.

**Comment by dscho** [comment]
The `rsync://` protocol was deprecated a real long time ago.

As to including `rsync`, [my suggestion stands](https://github.com/git-for-windows/git/issues/347#issuecomment-136997693).

**Comment by vteivans** [comment]
Hey There!
I would also really wan't rsync included in git for windows, cause that really fixes working with symlinks in vagrant box on Windows.
Are there any other resources for compiling rsync with git SDK?

**Comment by dscho** [comment]
@vteivans sure: https://github.com/git-for-windows/MSYS2-packages/tree/master/rsync and https://github.com/git-for-windows/git/wiki/Package-management (in particular, https://github.com/git-for-windows/git/wiki/Package-management#rebuild-packages).

**Comment by dscho** [comment]
(But I fear that you will be disappointed in hearing that rsync cannot fix whatever problems you have with symlinks)

**Comment by vteivans** [comment]
@dscho,  no I already fixed everything with Rsync, and it works like a charm, the only problem is setting it up, so it would be great if it would be included in git for windows. :)

**Comment by dscho** [comment]
@vteivans The Git for Windows project is about empowering users to bring about what they want, without the need to rely on other people.

So if you are dreaming of a Git for Windows that ships `rsync.exe`, you can continue dreaming while preparing a Pull Request that imitates https://github.com/git-for-windows/build-extra/commit/920c19bd0381a89735326eab2490fd27e1ca4951, including a verbose commit message that describes that Git has not been relying on rsync for eons but that it is still nice to have and that it only adds <number> kilobytes to the installer.

**Comment by geneorama** [comment]
I recently updated git from the git-scm site, and I have `rsync` now.  So, was it included and this thread can be closed?

**Comment by dscho** [comment]
@geneorama It was not included:

```sh
$ which rsync
which: no rsync in ([the entire PATH])
```

(this is with Git for Windows 2.7.4's Git Bash)

Even worse, Git's support for the `rsync` protocol will be officially dropped with 2.8.0. 

If there had been more action and less talking in this here ticket, I would actually have accepted a Pull Request to add `rsync.exe` to the official Git for Windows installer, but nothing happened at all, except for upstream removing rsync support officially. So now I am much less inclined to include a tool that turns out to be totally useless within Git's context.

Seriously, if you want to be able to install Unixy tools in your environment on Windows, Git for Windows (whose mission it is to bring *Git* to the Windows masses) is a suboptimal choice. You definitely want to go with [MSYS2](https://msys2.github.io/) or [Cygwin](https://cygwin.com/) instead.

And this ticket is already closed ;-)

**Comment by geneorama** [comment]
@dscho Your comment helped me considerably.  I realized that my version of rsync is indeed not from `git for windows`. I hadn't run `which rsync`.  Yes, insert forehead slap, eye rolling, and other nonverbal  chiding here.
My version is from [Rtools](https://cran.r-project.org/bin/windows/Rtools/).  The weird thing is that before updating git for windows the other day I simply got "rsync not found". I thought that updating _git for windows_ installed `rsync`.  Actually, I must have had an error in my PATH variable, and after updating _git for windows_ I happened to adjust my PATH.  It was this coincidental adjustment that made `rsync` available.

**Comment by dlong500** [comment]
Just thought I'd post here for others searching to let people know that at least as of now (9/7/2017) it is extremely easy to manually add rsync support to the Git for Windows environment without even having to resort to running the Git for Windows SDK.

I stumbled across [this post](https://blog.tiger-workshop.com/add-rsync-to-git-bash-for-windows/) and didn't think it could really be that easy given all the alternatives, but it really is.  All you need to do is grab the rsync package from the [MSYS2 Package repository](http://repo.msys2.org/msys/x86_64/) and extract the `rsync.exe` file into `C:\Program Files\Git\usr\bin`.  One and done.  It works.

If you want to also use Pageant authentication with rsync over ssh, the ssh-pageant binary is _already installed in Git for Windows_ and all you need to do (as listed in the [ssh-pageant usage section](https://github.com/cuviper/ssh-pageant)) is add a small bit to your `~/.bashrc` or `~/.bash_profile` file.  I've modified it specifically for use with the Git for Windows environment:
```
# ssh-pageant
eval $("/c/Program Files/Git/usr/bin/ssh-pageant" -r -a "/c/Windows/Temp/.ssh-pageant-$USERNAME")
```
I can't speak to how other utilities might or might not work by manually installing them this way, but rsync appears to work just fine.

**Comment by landstander668** [comment]
@dlong500 Pagent configuration is actually documented in the Git for Windows wiki [here](https://github.com/git-for-windows/git/wiki/OpenSSH-Integration-with-Pageant).

**Comment by dlong500** [comment]
@landstander668 Thanks for the link to the GfW specific ssh-pageant configuration info.  I'll review that and may update my post with a better ssh-pageant command (though my original appears to work just fine, albeit with more hard-coded paths).

**Comment by dlong500** [comment]
@landstander668 I've updated my above post with more robust ssh-pageant usage based on the info.  However, I've actually enhanced it some further because the commands on the wiki don't account for a `SSH_AUTH_SOCK` environment variable that is defined but points to a file that doesn't exist anymore (as happens to be the case on my machine and I suspect quite a few others as well).

**Comment by c33s** [comment]
i would love to see this ticket reopend. of course rsync is not required for git but git for windows really makes windows rock. together with http://cmder.net/ (which also bundles with git for windows) it unixify the windows environment. 

why i would prefer to have `rsync` directly in git for windows is because of the incompatibilities between ssh and rsync from different provider, see https://github.com/git-for-windows/git/issues/1401

having this in here allows us to use rsync which is updated regulary and perfectly work together with all the other shiped binaries.

**Comment by dscho** [comment]
> i would love to see this ticket reopend.

I have a *much* better idea: work on it. Use https://github.com/git-for-windows/build-extra/commit/16d8d828be9c0e5bc8b180c941388cf114e742fb as an example, including the information in the commit message.

That idea is much better because:
1. you have a much better chance of getting what you want, and
2. you avoid dumping even more work on the already overwhelmed Git for Windows maintainer.

**Comment by dscho** [comment]
> Use git-for-windows/build-extra@16d8d82 as an example

You could follow the link to the PR in this page, and then to the ticket linked from the PR, to get to the instructions I wrote how to get started with adding stuff to the installer. But I'll make it easy for you (having followed that trail already): https://github.com/git-for-windows/git/issues/1224#issuecomment-313948801

**Comment by cgilles** [comment]
Hi all,

I vote also to reopen this file as rsync will be a great addition to git-windows. I personally add rsync to my git-windows install manually following this tutorial, and it work as expected:

https://prasaz.medium.com/add-rsync-to-windows-git-bash-f42736bae1b3

Thanks in advance

Gilles Caulier

**Comment by dscho** [comment]
The purpose of [Git for Windows](https://git-for-windows.github.io/) is really to bring *Git* to Windows.

While that implies that we have to ship quite a few Unix-like tools, it does not mean that we will include such tools unless they are needed for Git's own operations, or at least for historical reasons (needs of active contributors are sometimes a factor in deciding whether to include a tool or not).

Having said that, what you really are looking for is [MSYS2](https://msys2.github.io/). Git for Windows leverages MSYS2 and ships with a subset of its files. MSYS2 even sports a package management system called "pacman" to install more tools (including Git...), so why don't you give it a whirl?

**Comment by geneorama** [comment]
@dscho Thanks for the advice and link, I will definitely consider it if I switch. 

I like Git Bash for Windows despite this flaw. I have chosen to assume that it was a thoughtful design decision or that nobody has been ambitious enough to solve it. I have managed to work around it by downloading rsync from somewhere and putting it on my path.

**Comment by dscho** [comment]
> I like Git Bash for Windows despite this flaw.

I vehemently reject the characterisation as "flaw" that I refuse to include `rsync` in Git for Windows for the benefit of only a handful people when _Git does not need it_.

**Comment by dlong500** [comment]
It certainly is not a flaw.  And furthermore it is relatively simple to manually add other packages to Git for Windows without requiring a full MSYS2 install.  I wrote up a an answer on ServerFault 6 years ago that I recently updated to account for some changes in packaging requirements (more updated than what I posted in this thread years ago).  It's [here](https://serverfault.com/questions/310337/using-rsync-from-msysgit-for-binary-files/872557#872557) if anyone might find it useful.
